proposals/record-structs.md

```diff
@@ -0,0 +1,301 @@
+
+    protected virtual bool PrintMembers(StringBuilder builder)
+    {
+        builder.Append(nameof(P1));
+        builder.Append(" = ");
+        builder.Append(this.P1); // or builder.Append(this.P1); if P1 has a value type
+        builder.Append(", ");
+
+        builder.Append(nameof(P2));
+        builder.Append(" = ");
+        builder.Append(this.P2); // or builder.Append(this.P2); if P2 has a value type
+
+        return true;
+    }
+
+    public override string ToString()
+    {
+        var builder = new StringBuilder();
+        builder.Append(nameof(R1));
+        builder.Append(" { ");
+
+        if (PrintMembers(builder))
+            builder.Append(" ");
+
+        builder.Append("}");
+        return builder.ToString();
+    }
+}
+```
+
+## Positional record struct members
+
+In addition to the above members, record structs with a parameter list ("positional records") synthesize
+additional members with the same conditions as the members above.
+
+### Primary Constructor
+
+A record struct has a public constructor whose signature corresponds to the value parameters of the
+type declaration. This is called the primary constructor for the type. It is an error to have a primary
+constructor and a constructor with the same signature already present in the struct.
+
+A record struct is not permitted to declare a parameterless primary constructor.
+
+Instance field declarations for a record struct are permitted to include variable initializers when there is a primary constructor. 
+At runtime the primary constructor executes the instance initializers appearing in the record-struct-body.
+
+If a record struct has a primary constructor, any user-defined constructor, except "copy constructor" must have an
+explicit `this` constructor initializer.
+
+Parameters of the primary constructor as well as members of the record struct are in scope within initializers of instance fields or properties. 
+Instance members would be an error in these locations (similar to how instance members are in scope in regular constructor initializers
+today, but an error to use), but the parameters of the primary constructor would be in scope and useable and
+would shadow members. Static members would also be useable.
+
+A warning is produced if a parameter of the primary constructor is not read.
+
+The definite assigment rules for struct instance constructors apply to the primary constructor of record structs. For instance, the following
+is an error:
+
+```csharp
+record struct Pos(int X) // definite assignment error in primary constructor
+{
+    private int x;
+    public int X { get { return x; } set { x = value; } } = X;
+}
+```
+
+### Properties
+
+For each record struct parameter of a record struct declaration there is a corresponding public property
+member whose name and type are taken from the value parameter declaration.
+
+For a record struct:
+
+* A public `get` and `init` auto-property is created (see separate `init` accessor specification).
```

Comment:
We did not decide on `init` properties, we decided on `set` properties: https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-10-05.md#record-struct-primary-constructor-defaults.

Reply:
Thanks. I'd missed that.

---
In reply to: [553003576](https://github.com/dotnet/csharplang/pull/4307#discussion_r553003576) [](ancestors = 553003576)